Add `{ignore,needs}-{rustc,std}-debug-assertions` directive support

jieyouxu opened the pull request:
Add `{ignore,needs}-{rustc,std}-debug-assertions` compiletest directives and retire the old `{ignore,only}-debug` directives. The old `{ignore,only}-debug` directives were ambiguous because you could have std built with debug assertions but rustc not built with debug assertions or vice versa. If we want to support the use case of controlling test run based on if rustc was built with debug assertions, then having `{ignore,only}-debug` will be very confusing.

cc @matthiaskrgr 

Closes #123987.

r? bootstrap (or compiler tbh)

jieyouxu committed:
compiletest: add `{ignore,needs}-{rustc,std}-debug-assertions` directive support

And retire the old `only-debug` directive which was ambiguous and only
for std debug assertions.

jieyouxu committed:
tests: `ignore-debug` -> `ignore-std-debug-assertions`

Mark-Simulacrum commented on tests/codegen/mem-replace-simple-type.rs:
```diff
@@ -1,6 +1,7 @@
 //@ compile-flags: -O -C no-prepopulate-passes
 //@ only-x86_64 (to not worry about usize differing)
-//@ ignore-debug: precondition checks make mem::replace not a candidate for MIR inlining
```
Just to check: we have some lint or w/e that will fail CI if someone adds a test with `ignore-debug` as we merge this PR?

jieyouxu replied:
compiletest (currently) have an allowlist of known directives, I've removed ignore-debug from that so compiletest will error if someone writes ignore-debug.

jieyouxu replied:
See https://github.com/rust-lang/rust/pull/131913#discussion_r1807864577

jieyouxu replied:
It will fail looking like (I know it's not super user friendly)

```
Testing stage1 compiletest suite=ui mode=ui (x86_64-pc-windows-msvc)
error: detected unknown compiletest test directive `only-debug` in X:\repos\rust\tests\ui\print_type_sizes\niche-filling.rs:13
errors encountered during EarlyProps parsing: X:\repos\rust\tests\ui\print_type_sizes\niche-filling.rs
thread 'main' panicked at src\tools\compiletest\src\header.rs:68:13:
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Mark-Simulacrum commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -1928,9 +1928,13 @@ NOTE: if you're sure you want to do this, please open an issue as to why. In the
 
         cmd.arg("--json");
 
+        if builder.config.rust_debug_assertions {
```
nit: can we rename this to rust_debug_assertions_rustc? Or maybe just rustc_debug_assertions and std_debug_assertions for the other one?

jieyouxu replied:
Yes, we can rename them to look less similar in prefix.

jieyouxu commented on src/tools/compiletest/src/command-list.rs:
```diff
@@ -45,7 +45,8 @@ const KNOWN_DIRECTIVE_NAMES: &[&str] = &[
     "ignore-coverage-run",
     "ignore-cross-compile",
-    "ignore-debug",
+    "ignore-debug-assertions-rustc",
```
Specifically here

jieyouxu commented:
Accompanying rustc-dev-guide PR: https://github.com/rust-lang/rustc-dev-guide/pull/2101

jieyouxu commented:
### Changes since last review

- Renamed compiletest/bootstrap fields/variables/flags to use the prefix form of `{rustc,std}-debug-assertions`. This does not change the `config.toml` option names of `debug-assertions` and `debug-assertions-std`.
- Added a directives unit test for the `{needs,ignore}-{rustc,std}-debug-assertions` to check that they are recognized.
- Updated the tests to use `ignore-std-debug-assertions`.

jieyouxu commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #131988) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu committed:
bootstrap: pass `--with-{rustc,std}-debug-assertions` to compiletest

And rename local variables and field names in bootstrap from
`debug_assertions{,_std}` -> `{rustc,std}-debug-assertions` to avoid
confusion where applicable.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11437717262/job/31817820102) [(plain)](https://github.com/rust-lang/rust/commit/c46a1ca8f3a9868aadea6e7a6395acb942a44bac/checks/31817820102/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
...i...ii.................

failures:

---- [codegen] tests/codegen/issues/issue-101082.rs#x86-64-v3 stdout ----

error in revision `x86-64-v3`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-101082.x86-64-v3/issue-101082.ll" "/checkout/tests/codegen/issues/issue-101082.rs" "--check-prefix=CHECK" "--check-prefix" "x86-64-v3" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-101082.rs:13:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-101082.rs:13:12: error: CHECK: expected string not found in input
 // CHECK: ret {{i64|i32}} 165
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-101082.x86-64-v3/issue-101082.ll:7:26: note: scanning from here
define noundef i64 @test() unnamed_addr #0 personality ptr @rust_eh_personality {
                         ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-101082.x86-64-v3/issue-101082.ll:11:66: note: possible intended match here
 %_3.sroa.5.0.iter.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 16


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-101082.x86-64-v3/issue-101082.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'issue_101082.775d9ed26ae14e6e-cgu.0' 
            2: source_filename = "issue_101082.775d9ed26ae14e6e-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: nofree norecurse nosync nounwind nonlazybind memory(none) uwtable 
            7: define noundef i64 @test() unnamed_addr #0 personality ptr @rust_eh_personality { 
check:13'0                              X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            8: bb4.preheader: 
check:13'0     ~~~~~~~~~~~~~~~
            9:  %iter = alloca [64 x i8], align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  call void @llvm.lifetime.start.p0(i64 64, ptr nonnull %iter) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  %_3.sroa.5.0.iter.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 16 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'1                                                                      ?          possible intended match
           12:  store i64 23, ptr %_3.sroa.5.0.iter.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           13:  %_3.sroa.5.sroa.4.0._3.sroa.5.0.iter.sroa_idx.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 24 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  store i64 16, ptr %_3.sroa.5.sroa.4.0._3.sroa.5.0.iter.sroa_idx.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  %_3.sroa.5.sroa.5.0._3.sroa.5.0.iter.sroa_idx.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 32 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  store i64 54, ptr %_3.sroa.5.sroa.5.0._3.sroa.5.0.iter.sroa_idx.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  %_3.sroa.5.sroa.6.0._3.sroa.5.0.iter.sroa_idx.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 40 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18:  store i64 3, ptr %_3.sroa.5.sroa.6.0._3.sroa.5.0.iter.sroa_idx.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19:  %_3.sroa.5.sroa.7.0._3.sroa.5.0.iter.sroa_idx.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 48 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  store i64 60, ptr %_3.sroa.5.sroa.7.0._3.sroa.5.0.iter.sroa_idx.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  %_3.sroa.5.sroa.8.0._3.sroa.5.0.iter.sroa_idx.sroa_idx = getelementptr inbounds i8, ptr %iter, i64 56 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22:  store i64 9, ptr %_3.sroa.5.sroa.8.0._3.sroa.5.0.iter.sroa_idx.sroa_idx, align 8 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           23:  %unmaskedload = load <4 x i64>, ptr %_3.sroa.5.0.iter.sroa_idx, align 8, !alias.scope !3 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  %0 = getelementptr inbounds i8, ptr %iter, i64 48 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           25:  %wide.masked.load.1 = call <4 x i64> @llvm.masked.load.v4i64.p0(ptr nonnull %0, i32 8, <4 x i1> <i1 true, i1 true, i1 false, i1 false>, <4 x i64> poison), !alias.scope !3 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           26:  %1 = add <4 x i64> %wide.masked.load.1, %unmaskedload 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27:  %2 = shufflevector <4 x i64> %1, <4 x i64> %unmaskedload, <4 x i32> <i32 0, i32 1, i32 6, i32 7> 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28:  %3 = tail call i64 @llvm.vector.reduce.add.v4i64(<4 x i64> %2) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29:  call void @llvm.lifetime.end.p0(i64 64, ptr nonnull %iter) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  ret i64 %3 
check:13'0     ~~~~~~~~~~~~
           31: } 
check:13'0     ~~
check:13'0     ~
check:13'0     ~
           33: ; Function Attrs: nounwind nonlazybind uwtable 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34: declare noundef i32 @rust_eh_personality(i32 noundef, i32 noundef, i64 noundef, ptr noundef, ptr noundef) unnamed_addr #1 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           36: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           37: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #2 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           39: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           40: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #2 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           42: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: read) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43: declare <4 x i64> @llvm.masked.load.v4i64.p0(ptr nocapture, i32 immarg, <4 x i1>, <4 x i64>) #3 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           45: ; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none) 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: declare i64 @llvm.vector.reduce.add.v4i64(<4 x i64>) #4 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           48: attributes #0 = { nofree norecurse nosync nounwind nonlazybind memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64-v3" } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: attributes #1 = { nounwind nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64-v3" } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50: attributes #2 = { mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51: attributes #3 = { nocallback nofree nosync nounwind willreturn memory(argmem: read) } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52: attributes #4 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           54: !llvm.module.flags = !{!0, !1} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           55: !llvm.ident = !{!2} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~
check:13'0     ~
check:13'0     ~
           57: !0 = !{i32 8, !"PIC Level", i32 2} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59: !2 = !{!"rustc version 1.84.0-nightly (e4ff4b3d7 2024-10-21)"} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60: !3 = !{!4} 
check:13'0     ~~~~~~~~~~~
           61: !4 = distinct !{!4, !5, !"_ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h67f110d36c636f02E: %self"} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           62: !5 = distinct !{!5, !"_ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h67f110d36c636f02E"} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

jieyouxu commented:
Rebased and split compiletest/bootstrap changes into two commits, but no functional changes.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #132317) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu commented:
Rebased, no changes.

jieyouxu commented:
r? bootstrap

onur-ozkan reviewed (approved):
LGTM!

Only one question:

jieyouxu commented on tests/codegen/mem-replace-big-type.rs:
```diff
@@ -4,7 +4,8 @@
 
 //@ compile-flags: -C no-prepopulate-passes -Zinline-mir=no
-//@ ignore-debug: precondition checks in ptr::read make them a bad candidate for MIR inlining
+//@ ignore-std-debug-assertions
+// Reason: precondition checks in ptr::read make them a bad candidate for MIR inlining
```
Oh it definitely *can* be, but it gets quite long and I found it a bit easier to read on a different line.

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit 0d5cc8ee96c0bf7fa041727891839da4ac57c042 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 0d5cc8ee96c0bf7fa041727891839da4ac57c042 -->
<!-- homu: {"type":"Approved","sha":"0d5cc8ee96c0bf7fa041727891839da4ac57c042","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
